<launch>
    <!-- 基本參數 -->
    <arg name="video_device" default="/dev/video6"/>
    <arg name="model" default="vitl"/>
    <arg name="model_path" default="$(find depth_anything_ros)/trained_data/depth_anything_v2_metric_hypersim_$(arg model).engine"/>
    <arg name="depth_scale" default="0.5"/>
    <arg name="use_rainbow_colormap" default="true"/>

    <!-- 相機參數 -->
    <arg name="camera_width" default="640"/>
    <arg name="camera_height" default="480"/>
    <arg name="camera_fps" default="30"/>

    <!-- TF 轉換參數 -->
    <arg name="base_roll" default="0"/>
    <arg name="base_pitch" default="0"/>
    <arg name="base_yaw" default="0"/>
    <arg name="camera_roll" default="1.5708"/>
    <arg name="camera_pitch" default="1.5708"/>
    <arg name="camera_yaw" default="0"/>

    <!-- 定義相機話題 -->
    <arg name="image_topic" value="/usb_cam/image_raw"/>
    <arg name="camera_info_topic" value="/usb_cam/camera_info"/>

    <!-- USB 相機節點 -->
    <node name="usb_cam" pkg="usb_cam" type="usb_cam_node" output="screen" respawn="true">
        <param name="video_device" value="$(arg video_device)"/>
        <param name="image_width" value="$(arg camera_width)"/>
        <param name="image_height" value="$(arg camera_height)"/>
        <param name="pixel_format" value="yuyv"/>
        <param name="camera_frame_id" value="usb_cam"/>
        <param name="camera_name" value="usb_cam"/>
        <param name="camera_info_url" value="file://$(find depth_anything_ros)/config/camera_calibration.yaml"/>
        <param name="framerate" value="$(arg camera_fps)"/>
        <param name="io_method" value="mmap"/>
        <!-- 相機影像參數 -->
        <param name="auto_exposure" value="false"/>
        <param name="exposure" value="100"/>
        <param name="brightness" value="100"/>
        <param name="contrast" value="100"/>
        <param name="saturation" value="100"/>
        <param name="sharpness" value="100"/>
        <param name="autofocus" value="false"/>
        <param name="focus" value="0"/>
    </node>

    <!-- 加載相機標定文件 -->
    <rosparam command="load" file="$(find depth_anything_ros)/config/camera_calibration.yaml"/>

    <!-- TF 轉換設置 -->
    <!-- map 到 base_link 的轉換 -->
    <node pkg="tf2_ros" type="static_transform_publisher" 
          name="map_to_base" 
          args="0 0 0 $(arg base_roll) $(arg base_pitch) $(arg base_yaw) map base_link"/>

    <!-- base_link 到 camera_link 的轉換 -->
    <node pkg="tf2_ros" type="static_transform_publisher" 
          name="base_to_camera_link" 
          args="0 0 0 $(arg camera_roll) $(arg camera_pitch) $(arg camera_yaw) base_link camera_link"/>

    <!-- camera_link 到 usb_cam 的轉換 -->
    <node pkg="tf2_ros" type="static_transform_publisher" 
          name="camera_to_optical" 
          args="0 0 0 -1.5708 0 -1.5708 camera_link usb_cam"/>

    <group ns="depth_anything">
        <!-- Nodelet manager -->
        <node pkg="nodelet" type="nodelet" name="nodelet_manager" args="manager"
              output="screen" respawn="true">
            <param name="num_worker_threads" value="4"/>
        </node>

        <!-- Depth estimation node -->
        <node name="depth_estimation" pkg="depth_anything_ros" type="depth_estimation_node" output="screen">
            <remap from="~input_image" to="$(arg image_topic)"/>
            <remap from="~depth_registered/image_rect" to="depth_registered/image_rect"/>
            <param name="model_path" value="$(arg model_path)"/>
            <param name="depth_scale" value="$(arg depth_scale)"/>
            <param name="use_rainbow_colormap" value="$(arg use_rainbow_colormap)"/>
            <!-- 效能優化參數 -->
            <param name="skip_frames" value="2"/>
            <param name="use_gpu_optimization" value="true"/>
            <param name="gpu_memory_fraction" value="0.7"/>
            <param name="enable_tensorrt" value="true"/>
        </node>

        <!-- Point cloud generation -->
        <node pkg="nodelet" type="nodelet" name="point_cloud_xyzrgb"
              args="load depth_image_proc/point_cloud_xyzrgb nodelet_manager" 
              output="screen">
            <remap from="rgb/camera_info" to="$(arg camera_info_topic)"/>
            <remap from="rgb/image_rect_color" to="$(arg image_topic)"/>
            <remap from="depth_registered/image_rect" to="depth_registered/image_rect"/>
            <remap from="depth_registered/points" to="depth_registered/points"/>
            
            <!-- 點雲生成參數 -->
            <param name="queue_size" value="10"/>
            <param name="approximate_sync" value="true"/>
            <param name="approximate_sync_max_interval" value="0.1"/>
            <param name="frame_id" value="map"/>
            <param name="use_exact_sync" value="false"/>
        </node>
    </group>

    <!-- 深度視覺化節點 -->
    <node name="depth_visualizer" pkg="depth_anything_ros" type="depth_visualizer_node" output="screen">
        <param name="depth_scale" value="0.5"/>
        <param name="max_depth" value="4.0"/>
        <param name="x_width" value="500"/>
        <param name="z_height" value="300"/>
        <param name="sample_step" value="4"/>
        <param name="fx" value="500.0"/>
        <param name="fy" value="500.0"/>
        <param name="cx" value="320.0"/>
        <param name="cy" value="240.0"/>
    </node>

    <!-- RViz -->
    <node type="rviz" name="rviz" pkg="rviz" args="-d $(find depth_anything_ros)/rviz/depth_anything.rviz">
        <param name="frame_id" value="map"/>
    </node>

</launch>
